File: detox/ios/DetoxTestRunner/Utilities/XCUIElement+ExtendedText.m
//
//  XCUIElement+ExtendedText.m
//  DetoxTestRunner
//
//  Created by Leo Natan (Wix) on 2/21/20.
//

#import "XCUIElement+ExtendedText.h"
#import "DTXTestCaseAssertions.h"

@implementation XCUIElement (ExtendedText)

- (void)dtx_clearText
{
	if(self.elementType != XCUIElementTypeTextField &&
	   self.elementType != XCUIElementTypeSecureTextField &&
	   self.elementType != XCUIElementTypeSearchField &&
	   self.elementType != XCUIElementTypeTextView)
	{
		DTXFail(@"%s is not valid for %@.", __FUNCTION__, self);
		return;
	}
	
	NSString* currentValue = self.value;
	if([currentValue isKindOfClass:NSString.class] == NO)
	{
		DTXFail(@"Tried to clear and enter text into a non string value");
		return;
	}
	
	NSString* clearTypeText = [@"" stringByPaddingToLength:currentValue.length withString:XCUIKeyboardKeyDelete startingAtIndex:0];
	[self typeText:clearTypeText];
}

- (void)dtx_replaceText:(NSString*)text
{
	[self dtx_clearText];
	
	[self typeText:text];
}

@end

File: detox/ios/DetoxTestRunner/Utilities/XCUIElementExtensions/XCUIElement+ExtendedText.swift
//
//  XCUIElement+ExtendedText.swift
//  DetoxTestRunner
//
//  Created by Alon Haiut on 10/10/2021.
//  Copyright © 2021 Wix. All rights reserved.
//

import Foundation
import XCTest

extension XCUIElement {
	
	func dtx_clearText() -> Void {
		
		if(self.elementType != .textField &&
			self.elementType != .secureTextField &&
			self.elementType != .searchField &&
			self.elementType != .textView)
		{
//			DTXFail("%s is not valid for %@.", #function, self)
			return;
		}

		guard let currentValue = self.value as? String else {
//			DTXFail(@"Tried to clear and enter text into a non string value");
			return;
		}

		let clearTypeText = "".padding(toLength: currentValue.count, withPad: "", startingAt: 0)
		
//		NSString* clearTypeText = [@"" stringByPaddingToLength:currentValue.length withString:XCUIKeyboardKeyDelete startingAtIndex:0];
//		[self typeText:clearTypeText];
		
		self.typeText(clearTypeText)
	}
	
	func dtx_replaceText(_ text: String) -> Void {
		self.dtx_clearText()
		self.typeText(text)
	}
}
